Refs #36688 - Provide option to use wget for the new Register Host feature

goarsna opened the pull request:
* [x] I am familiar with the [contributing](https://github.com/theforeman/foreman-documentation/blob/master/CONTRIBUTING.md) guidelines.

Please cherry-pick my commits into:

* [ ] Foreman 3.7/Katello 4.9 (planned Satellite 6.14)
* [ ] Foreman 3.6/Katello 4.8
* [ ] Foreman 3.5/Katello 4.7 (Satellite 6.13)
* [ ] Foreman 3.4/Katello 4.6 (EL8 only)
* [ ] Foreman 3.3/Katello 4.5 on EL7 & EL8 (Satellite 6.12 on EL8 only; orcharhino 6.4/6.5 on EL8 only)
* [ ] Foreman 3.2/Katello 4.4 on EL7 & EL8
* [ ] Foreman 3.1/Katello 4.3 on EL7 & EL8 (Satellite 6.11 EL7/8; orcharhino 6.3 on EL7/8)
* We do not accept PRs for Foreman older than 3.1.

maximiliankolb commented on guides/common/modules/proc_registering-hosts.adoc:
```diff
@@ -40,6 +40,8 @@ endif::[]
 ifdef::katello,satellite,orcharhino[]
 . In the *Activation Keys* field, enter one or more activation keys to assign to hosts.
 endif::[]
+. Optional: Select the *Use wget* option, if you want to use `wget` to register hosts.
```
```suggestion
. Optional: Select the *Use wget* option if you want to use `wget` to register hosts.
```

maximiliankolb commented on guides/common/modules/proc_registering-hosts.adoc:
```diff
@@ -68,12 +70,26 @@ Then register the hosts with a secure `curl` command, such as:
 ----
 +
-The following is an example of the `curl` command with the `--insecure` option:
+If you chose to use `wget`, register the hosts with a secure `wget` command, such as:
```
Please swap the order; show the default command "curl" first, then "wget" second.

goarsna replied:
The diff is a little bit confusing here. The order is

1.  curl secure
2. wget secure
3. gurl insecure
4. wget insecure

maximiliankolb replied:
I guess first curl https and http; then wget https http. not 100% which order is best.

goarsna replied:
Agreed, I've switched the order

maximiliankolb reviewed (commented):
Rest LGTM with the assumption that it'll get merged into Foreman as is.

maximiliankolb commented:
FYI: Your PR probably needs to be rebased again and adjust this: https://github.com/theforeman/foreman-documentation/pull/2348#discussion_r1300190420

goarsna commented:
Thanks @maximiliankolb, I've rebased my branch and incorporated your feedback.

One thing I've noticed: At some points we were still talking about the host we want to register using singular and at some points using plural form. As the file has been renamed to singular form in PR [#2348](https://github.com/theforeman/foreman-documentation/pull/2348), I've converted the remaining passages.

maximiliankolb commented on guides/common/modules/proc_registering-a-host.adoc:
I'd go with "your host" instead of "the host".

maximiliankolb commented on guides/common/modules/proc_registering-a-host.adoc:
```diff
@@ -134,9 +150,9 @@ endif::[]
 ifdef::foreman-el,foreman-deb,katello[]
 . Optional: This step is for the Katello users only.
-If you register {RHEL} or CentOS hosts, in the *Activation Keys* field, enter one or more activation keys to assign to registered hosts.
+If you register a {RHEL} or CentOS host, in the *Activation Keys* field, enter one or more activation keys to assign to the registered host.
```
```suggestion
If you register your {EL} host, in the *Activation Keys* field, enter one or more activation keys to assign to the registered host.
```

goarsna replied:
Kept as discussed

mjivraja commented on guides/common/modules/proc_registering-a-host.adoc:
```diff
@@ -7,73 +7,89 @@ You can register a host by using registration templates and set up various integ
 +
-If an attacker, located in the network between {Project} and a host, fetches the CA file from the first insecure call, the attacker will be able to access the content of the API calls to and from the registered host and the JSON Web Tokens (JWT).
+If an attacker, located in the network between {Project} and the host, fetches the CA file from the first insecure call, the attacker will be able to access the content of the API calls to and from the registered host and the JSON Web Tokens (JWT).
 Therefore, if you have chosen to deploy SSH keys during registration, the attacker will be able to access the host using the SSH key.
 +
 Instead, you can manually copy and install the CA file on each host before registering the host.
```
```suggestion
Therefore, it is highly recommended to manually copy and install the CA file on each host before registering them instead of using SSH keys.
```
Not a part of this PR but this should be precise.

Lennonka replied:
This is an incorrect suggestion. The SSH keys are related to remote execution. The user can deploy them regardless of using or not using the insecure call. It just says that these SSH keys can be stolen if the call is insecure.

mjivraja replied:
How about we make it more specific `Therefore in this case,....` 

I mean I am not overlooking SSH. But, particular to this case only.

Lennonka replied:
But that doesn't help. It's already specific enough.

mjivraja replied:
Alright, we can disregard this 👍

Lennonka commented on guides/common/modules/proc_registering-a-host.adoc:
```diff
@@ -7,73 +7,89 @@ You can register a host by using registration templates and set up various integ
 ----
 +
-The following is an example of the `curl` command with the `--insecure` option:
+The following is an example of the insecure `curl` command with the `--insecure` option:
 +
 [options="nowrap", subs="+quotes,attributes"]
 ----
 # curl -sS --insecure https://{foreman-example-com}/register ...
 ----
```
```suggestion
The following is an example of the insecure `curl` command:
+
[options="nowrap", subs="+quotes,attributes"]
----
# curl -sS *--insecure* https://{foreman-example-com}/register ...
----
```

maximiliankolb replied:
I am unsure about highlighting parts of the command; this seems rather unusual. To reduce the overall length, maybe we can instead append a line such as 

````
+
If your host does not trust the {Project} SSL certificate, use the `X` option.
````

goarsna replied:
Thinking about the whole "Insecure" bullet point IMO it would also be a good idea to omit the whole command examples. Users should simply copy the generated command and not modify it manually (as it is really a complex command) and the command details could make users think, they have to enter / modify things manually.

So we could just replace the part beginning from `Then register your host with a secure `curl` command, such as:` by `Afterwards you can continue registering your hosts without the *Insecure* option.`

What do you think about?

Lennonka replied:
Good point. I agree.

maximiliankolb replied:
I would not duplicate all command examples. I'd rather mention that hosts have to trust the ProjectServer SSL cert, or, alternatively, add the suggestion as above: https://github.com/theforeman/foreman-documentation/pull/2376/files#r1331083168

Lennonka commented on guides/common/modules/proc_registering-a-host.adoc:
```diff
@@ -7,73 +7,89 @@ You can register a host by using registration templates and set up various integ
+# wget --no-verbose -O- https://{foreman-example-com}/register ...
+----
++
+The following is an example of the insecure `wget` command with the `--no-check-certificate` option:
++
+[options="nowrap", subs="+quotes,attributes"]
+----
+# wget --no-verbose -O- --no-check-certificate https://{foreman-example-com}/register ...
+----
```
```suggestion
The following is an example of the insecure `wget` command:
+
[options="nowrap", subs="+quotes,attributes"]
----
# wget --no-verbose -O- *--no-check-certificate* https://{foreman-example-com}/register ...
----
```

Lennonka reviewed (commented):
Perhaps we could emphasize the option that makes the command insecure in the code sample instead of explicitly mentioning it? Like this: See the comments below.

Lennonka commented on guides/common/modules/proc_registering-a-host.adoc:
```diff
@@ -7,73 +7,89 @@ You can register a host by using registration templates and set up various integ
 To do this, find where {Project} stores the CA file by navigating to *Administer* > *Settings* > *Authentication* and locating the value of the *SSL CA file* setting.
 +
-Copy the CA file to the `/etc/pki/ca-trust/source/anchors/` directory on hosts and enter the following commands:
+Copy the CA file to the `/etc/pki/ca-trust/source/anchors/` directory on each host and enter the following commands:
```
```suggestion
Copy the CA file to the `/etc/pki/ca-trust/source/anchors/` directory on your host and enter the following commands:
```

Lennonka reviewed (commented):
LGTM! Just a couple of details.

maximiliankolb commented:
@goarsna Please link to the Foreman issue if you have one. Example on the commit message format: https://github.com/theforeman/foreman-documentation/pull/2432/commits

goarsna commented:
Thanks to both of you for your input! 

Rebased, suqashed and linked to Foreman issue (hopefully I did that right) :)

maximiliankolb commented on guides/common/modules/proc_registering-a-host.adoc:
```diff
@@ -27,59 +27,75 @@ endif::[]
 ----
 +
-The following is an example of the `curl` command with the `--insecure` option:
+The following is an example of the insecure `curl` command with the `--insecure` option:
 +
 [options="nowrap", subs="+quotes,attributes"]
 ----
 # curl -sS --insecure https://{foreman-example-com}/register ...
 ----
```
```suggestion
+
If your host does not trust the SSL cerfiticate of your {ProjectServer}, add the `--insecure` option to the registration command.
```

Lennonka replied:
Yeah, but the user doesn't add the option manually, they select the option in the GUI.

maximiliankolb replied:
Yes, my bad. Then it'd be "...., select the *Insecure* checkbox". @goarsna

maximiliankolb commented on guides/common/modules/proc_registering-a-host.adoc:
```diff
@@ -27,59 +27,75 @@ endif::[]
-Hosts will use this CA file to connect to {Project} with all future calls making them secure.
+. Optional: If you want to use `wget` to register your host to {Project}, select the *Use wget* option.
+By default, {Project} generates a `curl` command.
+. Optional: If you want to make the first call insecure, select the *Insecure* option.
```
```suggestion
. Optional: If your host does not trust the SSL certificate of your {ProjectServer}, select the *Insecure* option.
```

maximiliankolb commented on guides/common/modules/proc_registering-a-host.adoc:
```diff
@@ -27,59 +27,75 @@ endif::[]
+. Optional: If you want to use `wget` to register your host to {Project}, select the *Use wget* option.
+By default, {Project} generates a `curl` command.
+. Optional: If you want to make the first call insecure, select the *Insecure* option.
+During this first call, your host downloads the CA file from {Project}.
```
```suggestion
During the first call, your host downloads the CA file from {Project}.
```

maximiliankolb commented on guides/common/modules/proc_registering-a-host.adoc:
```diff
@@ -27,59 +27,75 @@ endif::[]
+By default, {Project} generates a `curl` command.
+. Optional: If you want to make the first call insecure, select the *Insecure* option.
+During this first call, your host downloads the CA file from {Project}.
+Your host will use this CA file to connect to {Project} with all future calls making them secure.
```
```suggestion
Your host will use this CA file to connect to {ProjectServer} with all future calls.
```

maximiliankolb commented on guides/common/modules/proc_registering-a-host.adoc:
```diff
@@ -27,59 +27,75 @@ endif::[]
+----
+# wget --no-verbose -O- https://{foreman-example-com}/register ...
+----
++
+The following is an example of the insecure `wget` command with the `--no-check-certificate` option:
++
+[options="nowrap", subs="+quotes,attributes"]
+----
+# wget --no-verbose -O- --no-check-certificate https://{foreman-example-com}/register ...
+----
```
```suggestion
```

Apply this suggestion if we also drop the insecure example for `curl`.

maximiliankolb commented on guides/common/modules/proc_registering-a-host.adoc:
```diff
@@ -27,59 +27,75 @@ endif::[]
 # curl -sS https://{foreman-example-com}/register ...
 ----
 +
-The following is an example of the `curl` command with the `--insecure` option:
+The following is an example of the insecure `curl` command with the `--insecure` option:
 +
 [options="nowrap", subs="+quotes,attributes"]
 ----
 # curl -sS --insecure https://{foreman-example-com}/register ...
 ----
```
I would drop line 67-73/69-75 and not document the insecure example. Is that OK for you @Lennonka ?

Lennonka replied:
Yep.

maximiliankolb reviewed (commented):
Please rebase to master. @goarsna 

Left some comments; the only thing up for discussion is "do we want to documentation `curl --insecure`"?

Rest LGTM; but I would appreciate an "ACK" from someone else after everything is resolved.

maximiliankolb commented on guides/common/modules/proc_registering-a-host.adoc:
```diff
@@ -39,7 +39,7 @@ ifdef::katello,satellite,orcharhino[]
 . Optional: If you want to use `wget` to register your host to {Project}, select the *Use wget* option.
 By default, {Project} generates a `curl` command.
-. Optional: If you want to make the first call insecure, select the *Insecure* option.
+. Optional: If your host does not trust the SSL certificate of your {ProjectServer}, select the *Insecure* option.
```
```suggestion
. If your host does not trust the SSL certificate of your {ProjectServer}, select the *Insecure* option.
```

goarsna commented on guides/common/modules/proc_registering-a-host.adoc:
```diff
@@ -27,59 +27,68 @@ endif::[]
-. Optional: Select the *Insecure* option, if you want to make the first call insecure.
-During this first call, hosts download the CA file from {Project}.
-Hosts will use this CA file to connect to {Project} with all future calls making them secure.
+. Optional: If you want to use `wget` to register your host to {Project}, select the *Use wget* option.
```
```suggestion
. If you want to use `wget` to register your host to {Project}, select the *Use wget* option.
```

goarsna commented:
Thanks @Lennonka and @maximiliankolb for your work! I've incorporated all feedback and rebased my PR.

goarsna commented on guides/common/modules/proc_registering-a-host.adoc:
```diff
@@ -27,59 +27,61 @@ endif::[]
 ----
 +
-Then register the hosts with a secure `curl` command, such as:
+Then register your host with a secure `curl` command, such as:
 +
 [options="nowrap", subs="+quotes,attributes"]
 ----
 # curl -sS https://{foreman-example-com}/register ...
 ----
 +
```
What I'm still thinking about is if we should leave out also the secure code examples. This would also apply for the `wget` example below. What do you think @Lennonka and @maximiliankolb?
```suggestion
```

Lennonka replied:
I'd remove them.

maximiliankolb reviewed (approved):
LGTM

maximiliankolb commented:
Converted back to "draft" because the PR in foreman has not been merged yet.

goarsna commented:
I've sorted the changes and put them into three commits.

Lennonka commented:
Please, rebase.

goarsna commented:
Thanks, @Lennonka! Rebased, checked the changes again, and added an extra commit to remove a duplicate step for the `katello` build target.

maximiliankolb reviewed (approved):
I have reviewed all four commits: LGTM!

Any other input?

adamlazik1 reviewed (approved):
LGTM stylewise, you can go ahead and merge.

goarsna commented:
Thanks a lot to all of you for your reviews and your effort. And thanks for merging :)

nofaralfasi commented:
I think we should revert this change, especially considering that we are currently in stabilization week and the PR itself has not been merged yet.
